### PR TITLE
[Data] Extend API to enable passing sample weights via ray.dataset.to_tf (#45701)

### LINKED ISSUE #45701: [Data] Add API support for passing sample weights via ray.dataset.to_tf
### Description

Add support for ray.dataset.to_tf to pass sample weights

https://www.tensorflow.org/versions/r2.8/api_docs/python/tf/keras/Model#fit

API does not currently support the feature: 
https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.to_tf.html#ray-data-dataset-to-tf 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by jeffreyjeffreywang**
## Why are these changes needed?
`tensorflow.keras.model.fit` supports callers to pass in sample weights along with input dataset (https://www.tensorflow.org/versions/r2.8/api_docs/python/tf/keras/Model#fit). However, `ray.data.Dataset.to_tf()` does not support this feature. Currently, callers can only generate data instances comprised of features and labels.

This PR extends `ray.data.Dataset.to_tf()` to allow callers to specify additional metadata associated with data samples. Specifically, `additional_columns` is introduced as a forced kwargs to `ray.data.Dataset.to_tf()` and `ray.DataIterator.to_tf()`. When `additional_columns` is not specified, there is no change to the API behavior. By contrast, while `additional_columns` is provided, the APIs will yield `additional_metadata` along with `features` and `labels`. To leverage the functionality for passing sample weights along with sample features and labels in tensorflow, one can create `tf.data.Dataset` from `ray.data.Dataset.to_tf()` and specifying `additional_columns="weight"` as `weight` is the column for storing sample weights in `ray.data.Dataset`.

We do not explicitly name the new argument `sample_weight` or `weight` and do not limit its type to be `str` as there may be other metadata associated with each sample that we want to yield while iterating through the dataset.

We follow the heuristic of inferring `feature_type_spec` and `label_type_spec` when determining `additional_type_spec`. Existing tests are extended to validate when the APIs are invoked with and without the `additional_columns` argument to ensure existing callers won't fail to invoke `ray.data.Dataset.to_tf()` and `ray.DataIterator.to_tf()` as this new change rolls out.

## Related issue number
Resolves #45701 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -4059,13 +4059,15 @@ def to_tf(
         feature_columns: Union[str, List[str]],
         label_columns: Union[str, List[str]],
         *,
+        additional_columns: Union[str, List[str]] = None,
```
Comment (jeffreyjeffreywang) [inline python/ray/data/dataset.py:4062]: I'm not thrilled about the naming of this argument. Open for brainstorming 😄 

Even though the current use case for this argument is to include sample weights, I don't think naming it as `sample_weight` or `weight` is a good idea as there may be other metadata that we want incorporate/yield in the future.

Comment (jeffreyjeffreywang) [reply python/ray/data/dataset.py:4062]: I also deliberately allow `List[str]` as its type as we may want to associate more than one metadata to each (feature, label) pair.

File: python/ray/data/iterator.py [hunk 2]
```diff
@@ -676,13 +676,15 @@ def to_tf(
...
         prefetch_batches: int = 1,
         batch_size: int = 1,
         drop_last: bool = False,
         local_shuffle_buffer_size: Optional[int] = None,
         local_shuffle_seed: Optional[int] = None,
         feature_type_spec: Union["tf.TypeSpec", Dict[str, "tf.TypeSpec"]] = None,
         label_type_spec: Union["tf.TypeSpec", Dict[str, "tf.TypeSpec"]] = None,
+        additional_type_spec: Union["tf.TypeSpec", Dict[str, "tf.TypeSpec"]] = None,
     ) -> "tf.data.Dataset":
         """Return a TF Dataset over this dataset.
```
Comment (jeffreyjeffreywang) [inline python/ray/data/iterator.py:690]: Hey @scottjlee, let me know if you think adding an example specifically for explaining the usage of the new argument is a good idea.

File: python/ray/data/tests/test_tf.py [hunk 3]
```diff
@@ -28,76 +28,179 @@ def test_autosharding_is_disabled(self):
         expected_auto_shard_policy = tf.data.experimental.AutoShardPolicy.OFF
         assert actual_auto_shard_policy is expected_auto_shard_policy
 
-    def test_element_spec_type(self):
-        ds = ray.data.from_items([{"spam": 0, "ham": 0}])
-
-        dataset = ds.to_tf(feature_columns="spam", label_columns="ham")
+    @pytest.mark.parametrize("include_additional_columns", [False, True])
```
Comment (jeffreyjeffreywang) [inline python/ray/data/tests/test_tf.py:31]: Note for reviewers: I intend to validate the API with and without (which shall fall back to the original behavior) the new arguments to ensure that the API is backward compatible. I understand there may be some redundancy in the following tests. Let me know if you think there is a better approach for testing.

**Review by scottjlee — COMMENTED** [review]
thanks for updating all the tests!

can we also add an additional unit test, where we use `Dataset.to_tf()` with sample weights passed, then we call `keras.Model.fit()` with the sample weights? so that we can verify that the original intended behavior

File: python/ray/data/iterator.py [hunk 4]
```diff
@@ -842,9 +861,25 @@ def generator():
...
+            )
+        else:
+            dataset = tf.data.Dataset.from_generator(
+                generator,
+                output_signature=(
+                    feature_type_spec,
+                    label_type_spec,
+                    additional_type_spec,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (jeffreyjeffreywang) [reply python/ray/data/iterator.py:882]: Hey @scottjlee, thank you so much for taking some time to go over my PR! I'll address your comments and add the test you mentioned above.

File: python/ray/data/iterator.py [hunk 5]
```diff
@@ -842,9 +861,25 @@ def generator():
             feature_type_spec = get_type_spec(schema, columns=feature_columns)
             label_type_spec = get_type_spec(schema, columns=label_columns)
 
-        dataset = tf.data.Dataset.from_generator(
-            generator, output_signature=(feature_type_spec, label_type_spec)
-        )
+        if additional_columns is not None and additional_type_spec is None:
+            schema = self.schema()
+            valid_columns = set(schema.names)
```
[the remark this answers is not in the document]
Comment (jeffreyjeffreywang) [reply python/ray/data/iterator.py:866]: `valid_columns` is used in `validate_columns()` to verify `additional_columns` refers to an existing column name. If we remove this line and `feature_type_spec` and `label_type_spec` are both not `None`, `valid_columns` will not be set on line 858. When we enter this `if` block on line 864, there will be no `valid_columns` to check against.

File: python/ray/data/iterator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffreyjeffreywang) [reply python/ray/data/iterator.py:882]: The test for using `Dataset.to_tf()` with sample weights passed and calling `keras.Model.fit()` with the sample weights is actually already included with TestToTF.test_training with `include_additional_columns=True`. @scottjlee , let me know if you think this test is sufficient to validate the original intended behavior.

**Comment by jeffreyjeffreywang** [comment]
@scottjlee Would appreciate your quick review when you have time, no rush 😄

File: python/ray/data/iterator.py (hunk 2 above)
Comment (scottjlee) [reply python/ray/data/iterator.py:692]: yes good idea, could you add an example that demonstrates how to use `feature_type_spec`, `label_type_spec`, and `additional_type_spec`? thanks!

File: python/ray/data/iterator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (scottjlee) [reply python/ray/data/iterator.py:868]: yes you're right, my mistake

File: python/ray/data/iterator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottjlee) [reply python/ray/data/iterator.py]: ah yeah i missed that, thanks! that should be sufficient.

**Review by scottjlee — APPROVED** [review]
thanks for the contribution!

File: python/ray/data/iterator.py (hunk 2 above)
Comment (jeffreyjeffreywang) [reply python/ray/data/iterator.py:692]: Yeah definitely, will do!

**Review by omatthew98 — APPROVED** [review]
Lgtm, thanks for the contribution!

**Comment by scottjlee** [comment]
@jeffreyjeffreywang let me know when the docs example is added, and i can take a final look before we get this merged

**Comment by jeffreyjeffreywang** [comment]
Hey @scottjlee, definitely, I plan to get it done today.

**Comment by jeffreyjeffreywang** [comment]
Hey @scottjlee, I've added some documentation to explain the spec arguments and the new argument I introduced. Doc examples are validated with `pytest --dodctest-modules` and built locally with `make develop` in the doc directory. Here are the sections/examples I added. Please let me know if these look good to you 😄 

### ray.data.Dataset.to_tf
Screenshot 2024-08-01 at 6 00 37 PM

### ray.data.DataIterator.to_tf
Screenshot 2024-08-01 at 6 01 59 PM

**Comment by scottjlee** [comment]
Thanks, the docs changes look good to me. I am merging in latest master to resolve the build issue (unrelated to this PR)

**Comment by jeffreyjeffreywang** [comment]
Thanks Scott! Looking forward to contributing more 😃
